### PR TITLE
[RBAC] Harden RBAC policies for the operator

### PR DESCRIPTION
**Pull request by surajssd**
Right now we ask user to create ClusterRole and ClusterRoleBinding.
Which grants the operator cluster wide permissions, which is too
much of permissions.

So this PR tries to limit those permissions by assigning only Role
and RoleBinding, which are namespace limited.

Fixes https://github.com/habitat-sh/habitat-operator/issues/338

### CODE REVIEW

**Comment by surajssd** [comment]
Few questions though:

1) Do we want to permanently limit this operator to work only in single namespace?
2) Or do we want it to switch between ClusterScope or Namespaced?
3) If 2 is true then right now I am checking that using flag `namespace` if value is given then the code scopes it to single namespace. But then it is expected from user to create appropriate `Roles`, `RoleBindings` and `CRD` definition.

**Comment by surajssd** [comment]
If you want to test this and create CRD manually then the definition looks like this:

```yaml
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: habitats.habitat.sh
spec:
  group: habitat.sh
  names:
    kind: Habitat
    listKind: HabitatList
    plural: habitats
    shortNames:
    - hab
    singular: habitat
  scope: Namespaced
  version: v1beta1
```

**Comment by surajssd** [comment]
@krnowak added those flags, as you mentioned. Now about testing this I have some questions.

IIUC, the test setup code starts the operator once and then test runs against it. Now deploy multiple operator should the code base of setting up operator change? WDYT?

File: cmd/habitat-operator/main.go [hunk 1]
```diff
@@ -45,10 +47,29 @@ type Clientsets struct {
...
 
+// FlagOpts struct is used to save all the flag values for operator
+type FlagOpts struct {
+	Namespace           string
+	AssumeCRDRegistered bool
+}
+
+// validateFlags tries to ensure no conflicting flags are given at input
+func validateFlags(f *FlagOpts) error {
+	if f.AssumeCRDRegistered && f.Namespace == v1.NamespaceAll {
```
Comment (krnowak) [inline cmd/habitat-operator/main.go]: `AssumeCRDRegistered` and `Namespace` should be orthogonal. You should be able to pass any combination of those flags. Currently only you can pass only `--assume-crd-registered --namespace=foo` or none of those at all.

Of course, passing only `--namespace=foo` will make the operator to require a ClusterRole to have permissions for getting and creating CRDs, but that's fine. Same with passing just `--assume-crd-registered` which will require a ClusterRole for permissions to list namespaces.

So, I think I would just drop this function.

File: pkg/controller/v1beta2/controller.go [hunk 2]
```diff
@@ -209,7 +210,17 @@ func (hc *HabitatController) Run(ctx context.Context, workers int) error {
 }
 
 func (hc *HabitatController) cacheHabitats() {
-	hc.habInformer = hc.config.HabitatInformerFactory.Habitat().V1beta1().Habitats().Informer()
```
Comment (krnowak) [inline pkg/controller/v1beta2/controller.go]: I don't know if this change should be necessary. Instead of creating informer that way, maybe you could just change the way we create the informer factory (the `HabitatInformerFactory` and `KubeInformerFactory` members).

For `KubeInformerFactory` we currently use `NewSharedInformerFactory`. You could use `NewFilteredSharedInformerFactory`, where you can pass a namespace. For the last parameter (`tweakListOptions` or something, you could just pass `nil`).

Same for `HabitatInformerFactory`.

vendor/k8s.io/client-go/informers/factory.go

File: pkg/controller/v1beta2/controller.go [hunk 3]
```diff
@@ -221,7 +232,17 @@ func (hc *HabitatController) cacheHabitats() {
 }
 
 func (hc *HabitatController) cacheConfigMaps() {
-	hc.cmInformer = hc.config.KubeInformerFactory.Core().V1().ConfigMaps().Informer()
```
Comment (krnowak) [inline pkg/controller/v1beta2/controller.go]: Likely unnecessary change if changing the way we create the informer factory will be enough.

File: pkg/controller/v1beta2/stateful_sets.go [hunk 4]
```diff
@@ -286,7 +286,17 @@ func (hc *HabitatController) newStatefulSet(h *habv1beta1.Habitat) (*appsv1.Stat
 }
 
 func (hc *HabitatController) cacheStatefulSets() {
-	hc.stsInformer = hc.config.KubeInformerFactory.Apps().V1().StatefulSets().Informer()
+	hc.stsInformer = cache.NewSharedIndexInformer(
```
Comment (krnowak) [inline pkg/controller/v1beta2/stateful_sets.go]: Likely unnecessary change if changing the way we create the informer factory will be enough.

File: test/e2e/v1beta1/resources/namespaced-standalone/habitat.yaml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+apiVersion: habitat.sh/v1beta1
+kind: Habitat
+metadata:
+  name: example-namespaced-habitat
```
Comment (krnowak) [inline test/e2e/v1beta1/resources/namespaced-standalone/habitat.yaml]: Is the namespaced operator working in a default namespace? Could you try using some other namespace "testing-namespaced-or-something"?

File: pkg/controller/v1beta2/controller.go (hunk 2 above)
Comment (surajssd) [reply pkg/controller/v1beta2/controller.go]: done!

File: pkg/controller/v1beta2/controller.go (hunk 3 above)
Comment (surajssd) [reply pkg/controller/v1beta2/controller.go]: done

File: pkg/controller/v1beta2/stateful_sets.go (hunk 4 above)
Comment (surajssd) [reply pkg/controller/v1beta2/stateful_sets.go]: done

File: test/e2e/v1beta1/resources/namespaced-standalone/habitat.yaml (hunk 5 above)
Comment (surajssd) [reply test/e2e/v1beta1/resources/namespaced-standalone/habitat.yaml]: yes works in separate namespace!

```console
$ # build image
$ make image
# Compile statically linked binary for linux.
GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -ldflags="-s" -o habitat-operator github.com/habitat-sh/habitat-operator/cmd/habitat-operator                                                                      
docker build -t "habitat/habitat-operator:v0.7.2-23-g2189cac7" .
Sending build context to Docker daemon 93.99 MB
Step 1/3 : FROM scratch
 --->
Step 2/3 : COPY habitat-operator /habitat-operator
 ---> d596478b62f3
Step 3/3 : ENTRYPOINT ["/habitat-operator", "-logtostderr"]
 ---> Running in 495ae029d462
Removing intermediate container 495ae029d462
 ---> 29c15f2b0c69
Successfully built 29c15f2b0c69
Successfully tagged habitat/habitat-operator:v0.7.2-23-g2189cac7


$ # create ns, sa, role and rolebinding
$ kubectl apply -f examples/rbac-restricted/rbac-restricted.yaml
namespace/habitat-operator created
serviceaccount/habitat-operator created
rolebinding.rbac.authorization.k8s.io/habitat-operator created
role.rbac.authorization.k8s.io/habitat-operator created


$ # change ns
$ kubectl config set-context $(kubectl config current-context) --namespace habitat-operator
Context "minikube" modified.


$ # verify changed ns
$ kubectl get sa default -o jsonpath='{.metadata.namespace}';
habitat-operator

$ # create crd
$ kubectl apply -f examples/rbac-restricted/crd.yaml
customresourcedefinition.apiextensions.k8s.io/habitats.habitat.sh configured


$ # create operator deployment
$ kubectl apply -f examples/rbac-restricted/habitat-operator.yaml
deployment.extensions/habitat-operator created


$ # operator running
$ kubectl get pods
NAME                                READY     STATUS    RESTARTS   AGE
habitat-operator-7bb6788fcd-sqm65   1/1       Running   0          5s

$ # create habitat resource
$ kubectl apply -f examples/standalone/habitat.yml 
habitat.habitat.sh/example-standalone-habitat created


$ # operator logs
$ kubectl logs habitat-operator-7bb6788fcd-sqm65 
W0828 07:59:19.852873       1 client_config.go:552] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
ts=2018-08-28T07:59:19.853986365Z level=info component=controller/v1beta2 msg="Watching Habitat objects"
ts=2018-08-28T07:59:44.789126117Z level=info component=controller/v1beta2 msg="created StatefulSet" name=example-standalone-habitat
ts=2018-08-28T07:59:44.820780171Z level=info component=controller/v1beta2 msg="Created peer IP ConfigMap" name=peer-watch-file
ts=2018-08-28T07:59:52.000322309Z level=info component=controller/v1beta2 msg="Updated peer IP in ConfigMap" name=peer-watch-file ip=172.17.0.7

$ # pods
$ kubectl get pods
NAME                                READY     STATUS    RESTARTS   AGE
example-standalone-habitat-0        1/1       Running   0          4m
habitat-operator-7bb6788fcd-sqm65   1/1       Running   0          4m
```

**Comment by surajssd** [comment]
Fixing tests right now.

File: Makefile [hunk 6]
```diff
@@ -60,8 +64,13 @@ update-version:
 	sed \
 		-i.bak \
 		-e "s/\(habitat-operator:v\).*/\1$$(cat VERSION)/g" \
-		test/e2e/v1beta1/resources/operator/deployment.yml
-	rm -f test/e2e/v1beta1/resources/operator/deployment.yml.bak
+		test/e2e/v1beta1/clusterwide/resources/operator/deployment.yml
+	rm -f test/e2e/v1beta1/clusterwide/resources/operator/deployment.yml.bak
```
Comment (krnowak) [inline Makefile]: Could you change it to
```
for f in test/e2e/v1beta1/clusterwide/resources/operator/deployment.yml test/e2e/v1beta1/namespaced/resources/operator/deployment.yaml; do
    …
done
```
?

Comment (krnowak) [reply Makefile]: Or maybe even `for f in test/e2e/v1beta1/{clusterwide, namespaced}/resources/operator/deployment.yml`.

File: Makefile [hunk 7]
```diff
@@ -31,17 +31,21 @@ e2e: clean-test
...
+	-kubectl delete namespace testing-clusterwide
 	-kubectl delete clusterrolebinding habitat-operator-v1beta1
 	-kubectl delete clusterrole habitat-operator-v1beta1
+	-kubectl delete crd habitats.habitat.sh
+	-kubectl delete namespace testing-namespaced
 
 .PHONY: update-version
 update-version:
-	find examples -name "*.yml" -type f \
+	find examples \( -name "*.yaml" -o -name "*.yml" \) -type f \
```
Comment (krnowak) [inline Makefile]: Could you use `yml` extension for now to avoid this complication?

At some point we need to decide which of `yaml` and `yml` to use and be consistent with it.

File: examples/rbac-restricted/crd.yaml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (krnowak) [inline examples/rbac-restricted/crd.yaml]: I think we need to have some CI script that checks that this CRD manifest and the one in tests are in sync. I suppose that a simple `diff` will suffice.

File: hack/rbac-check.sh [hunk 9]
```diff
@@ -78,3 +78,19 @@ then
 else
     exit 1
 fi
+
+# Checking the namespaced RBAC rules
```
Comment (krnowak) [inline hack/rbac-check.sh]: Both roles and clusterroles have a common part, so it would be nice to make sure that this common part is consistent across all the RBAC files we have in the tree. I'm not really sure how to do it, though. Either extend the script to be smarter or splitting the clusterrole files into two clusterroles files where one would contain only the common part and other one would contain only the CRD and namespaces permissions? Any other ideas?

File: Makefile [hunk 10]
```diff
@@ -31,17 +31,21 @@ e2e: clean-test
...
+	go test -v ./test/e2e/v1beta1/namespaced/... --image "$(TESTIMAGE)" --kubeconfig $(KUBECONFIG_PATH) --ip "$(IP)"
 
 .PHONY: clean-test
 clean-test:
-	-kubectl delete namespace testing-v1beta1
+	-kubectl delete namespace testing-clusterwide
 	-kubectl delete clusterrolebinding habitat-operator-v1beta1
 	-kubectl delete clusterrole habitat-operator-v1beta1
+	-kubectl delete crd habitats.habitat.sh
+	-kubectl delete namespace testing-namespaced
```
[the remark this answers is not in the document]
Comment (krnowak) [reply Makefile]: Nevermind, roles and rolebindings are namespaced resources.

File: Makefile (hunk 6 above)
Comment (surajssd) [reply Makefile]: done

File: Makefile (hunk 7 above)
Comment (surajssd) [reply Makefile]: done!

File: hack/rbac-check.sh (hunk 9 above)
Comment (surajssd) [reply hack/rbac-check.sh]: I feel this can be better written in golang uisng the structs and parsing it and then comparing rules in files. WDYT?

Comment (surajssd) [reply hack/rbac-check.sh]: Can I create another issue for it and then take it from there?

Comment (surajssd) [reply hack/rbac-check.sh]: @krnowak created another issue to carry this part forward https://github.com/habitat-sh/habitat-operator/issues/347

File: test/e2e/v1beta1/namespaced/main_test.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+		return errors.Wrap(err, "create Deployment failed")
+	}
+
+	if err := f.WaitForResources("name", d.ObjectMeta.Name, 1); err != nil {
+		return errors.Wrap(err, "waiting for the namespaced Operator deployment to be ready")
+	}
+	return nil
+}
+
+// createRBAC creates RBAC rules in the cluster necessary for the clusterwide
```
Comment (krnowak) [inline test/e2e/v1beta1/namespaced/main_test.go]: `s/clusterwide/namespaced/`

File: hack/crd-check.sh [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+set -euo pipefail
+
+function say
+{
+    printf '%s\n' "${1}"
+}
+
+# CRD file used in example and the test should be in sync.
+readonly example='examples/rbac-restricted/crd.yml'
```
[the remark this answers is not in the document]
Comment (surajssd) [reply hack/crd-check.sh]: done

File: helm/habitat-operator/templates/clusterrole.yaml [hunk 13]
```diff
@@ -51,3 +53,4 @@ rules:
   - events
   verbs: [ "create", "patch" ]
 {{- end }}
+{{- end }}
```
[the remark this answers is not in the document]
Comment (surajssd) [reply helm/habitat-operator/templates/clusterrole.yaml]: done

File: helm/habitat-operator/templates/crd.yaml [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  names:
+    kind: Habitat
+    listKind: HabitatList
+    plural: habitats
+    shortNames:
+    - hab
+    singular: habitat
+  scope: Namespaced
+  version: v1beta1
+{{- end }}
```
[the remark this answers is not in the document]
Comment (surajssd) [reply helm/habitat-operator/templates/crd.yaml]: done

File: helm/habitat-operator/templates/role.yaml [hunk 15]
```diff
@@ -0,0 +1,51 @@
+{{- if .Values.operatorNamespaced }}
+{{- if .Values.rbacEnable }}
+{{- if .Capabilities.APIVersions.Has "rbac.authorization.k8s.io/v1beta1" }}
```
[the remark this answers is not in the document]
Comment (surajssd) [reply helm/habitat-operator/templates/role.yaml]: sure

File: test/e2e/v1beta1/namespaced/main_test.go (hunk 11 above)
Comment (surajssd) [reply test/e2e/v1beta1/namespaced/main_test.go]: done

File: helm/habitat-operator/templates/role.yaml (hunk 15 above)
[the remark this answers is not in the document]
Comment (surajssd) [reply helm/habitat-operator/templates/role.yaml]: https://github.com/habitat-sh/habitat-operator/issues/352 created one!

File: cmd/habitat-operator/main.go [hunk 16]
```diff
@@ -26,8 +26,11 @@ import (
 
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
 	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (krnowak) [inline cmd/habitat-operator/main.go]: I'm surprised that golang does not warn that you are importing the same package twice under different names. Please drop the one with `v1` name and use the `metav1`.

Comment (surajssd) [reply cmd/habitat-operator/main.go]: done
